### DISCUSSION TITLE
Confirming Bounty Master Funds

### DISCUSSION THREAD

**Discussion by Hodlatoor** [ubiquity/ubiquibot#239] [Ideas]
Here is the situation I encountered:
Worked on a bounty. finished. went to claim and get an error message. The bounty master's wallet had insufficient funds.
big problem that should be exorcised.

one thing a bounty hunter should not worry about is getting paid on DevPool. Two thoughts on this:
1. Have a little icon/image (in green) next to each open bounty that says "Funds confirmed".  this way the bounty hunter knows he will be getting paid.
2.  Thoughts on how to manage on backend - each bounty master should deposit funds which are an account balance (keeping in mind gas fees consider it a floating/virtual balance. more on this later).
2.1 once a bounty master opens an issue the funds for the task are deducted from the balance and earmarked to the bounty (an escrow).  this cannot be removed or cancelled once a bounty hunter locks in the task. 
3. since gas fees will be a concern when moving fudns from bounty master to bounty master DevPool account and then accounted forward to developers. How about holding the funds in the DevPool and having a side registry which batches the transactions for each bounty movement when needed. 
OR
if we want the transaction to go from master to hunter directly, establishing "secured credit lines" for each bounty master by which they deposit a % to devpool, which they cannot exceed, for their projects until we see they are not going to rug. OR to avoid a long term mega rug- limit the number of open bounties and $ amounts.

**Comment by 0x4007** [comment]
> one thing a bounty hunter should not worry about is getting paid on DevPool.

Agreed

> 1.

Requires our Chrome Extension to inject a custom UI element on GitHub. Possible but backlogged.

> 2. 

Not sure how an escrow can be done on chain in a gas efficient manner (particularly for small bounties.) Perhaps there is a more viable architecture here @rndquu 

> 3. 

We have discussed the concept of "secured credit lines" aka insurance in the research meeting last week, in [this topic](https://ubiquitydao.notion.site/Bounty-Hunter-Protection-0ef930ed5ab648ec9fbe8af0e3fb4800), but deemed that it's not a good idea to get into insurance. We don't want to custody partner funds in order to limit our own liabilities. @Draeieg if you have more color to add on this topic from our conversation please share.

**Comment by Hodlatoor** [comment]
2. a chrome extension does not address the entirety of the market and does it not open up more potential attack vectors?

3. secured credit lines not as insurance. but think like a debit card. For eg. bounty master charges it up to $1000 and you can only setup up to $1000 in bounties.  I think if we don't address this there is concern for malicious actors taking advantage of the system and of hunters. both of which we want to avoid.

**Comment by rndquu** [comment]
We can simply restrict bounty hunters from getting assigned to an issue if there are not enough funds in the bot's wallet address to cover the bounty payment.

Example:
1. Bounty hunter posts `/assign`
2. The bot checks that its address has enough funds to cover the payout
3. If there are not enough funds then print smth like "Not enough funds on the bot's wallet address" else assign the issue as usual.

**Reply by Hodlatoor** [reply]
but that creates a FIFO issue for funds when a master issues more bounties than he has funds available.  best way to avoid this is confirmed (locked) funds.

Simplicity and transparency. If a hunter sees a bounty they should implicitly know funds are there. 100% of the times. Not to go read/hope/misunderstand/get excited and get let down.

**Reply by rndquu** [reply]
Bounty master can create as many issues as he wants but bounty hunters can self assign issues only while there are funds in the wallet. If a bounty hunter is unable to self assign an issue then something is wrong and he shouldn't even start it because he can't be sure that the bounty will be paid. 

It is much easier to implement this feature on the bot's side compared to some escrow smart contract for bounty masters.

**Reply by 0x4007** [reply]
I also want to add that the ethos of DeFi is about capital efficiency. Locking up capital guarantees security for the bounty hunters, I agree; but rndquu's proposal is leaner regarding capital lockup (considering if an organization generates a ton of bounties, which is desirable behavior) and solves the business problem.

**Reply by 0x4007** [reply]
@rndquu can you make your proposal into a bounty and link it here?

**Reply by rndquu** [reply]
> @rndquu can you make your proposal into a bounty and link it here?

https://github.com/ubiquity/bounty-bot/issues/264
